msal/authority.py

```diff
@@ -53,15 +53,20 @@ def __init__(self, authority_url, http_client, validate_authority=True):
             performed.
         """
         self._http_client = http_client
-        authority, self.instance, tenant = canonicalize(authority_url)
-        parts = authority.path.split('/')
-        is_b2c = any(self.instance.endswith("." + d) for d in WELL_KNOWN_B2C_HOSTS) or (
+        self.authority, self.instance, self.tenant = canonicalize(authority_url)
+        parts = self.authority.path.split('/')
+        self.is_b2c = any(self.instance.endswith("." + d) for d in WELL_KNOWN_B2C_HOSTS) or (
             len(parts) == 3 and parts[2].lower().startswith("b2c_"))
-        if (tenant != "adfs" and (not is_b2c) and validate_authority
+        self.authority_url = authority_url
+        self.validate_authority = validate_authority
+        self.is_initialized = False
+
+    def initialize(self):
```

Comment:
Also a possible trick here. We could define all the actual initialization work in a `__initialize()`. And then the public `initialize()` would just allow [reentrancy](https://en.wikipedia.org/wiki/Reentrancy_(computing)).

Once we reach that far, we can go back to revisit how the application.py would be able to call this.